### ISSUE TITLE
Only variable declarations and control directives are allowed inside functions

### ISSUE THREAD

**Issue by lmartins** [sass/libsass#1057]
Testing the beta release of gulp-sass, which uses node-sass 3.0.0-beta.4, I've started getting an error that didn't happen before:
```
[11:32:04] Starting 'sass'...
[11:32:05] Finished 'sass' after 1.27 s

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: bower_components/modular-scale/stylesheets/modular-scale/_sort-list.scss
  35:5  only variable declarations and control directives are allowed inside functions
```
Line 35 refers to the closing bracket of the following code:
```scss
    // Check if the A value is smaller
    // Uses or equal to avoid duplicate numbers
    @if nth($A, $A-counter) <= nth($B, $B-counter) {
      $Return: join($Return, nth($A, $A-counter));
      $A-counter: $A-counter + 1;
    }
```
After opening this on the [gulp-sass repo](https://github.com/dlmanning/gulp-sass/issues/231#issuecomment-90013364) I was instructed to report this here.

**Comment by xzyfer** [comment]
Hi @imathis 

Please provide a fully working example that produces the error you mention.

**Comment by Snugug** [comment]
@xzyfer This is actually a regression from `3.2.0-beta.2`. I just ran into this issue having upgraded to `3.2.0-beta.5`. I'll get you a reduced use case in a few.

**Comment by xzyfer** [comment]
Great thanks, without a test case we can't really begin fixing the issue :)

**Comment by Snugug** [comment]
Changing the [@elseif](https://github.com/modularscale/modularscale-sass/blob/2.x/stylesheets/modular-scale/_sort-list.scss#L38) to `@else if` resolves this issue, although raises other question.

First, how did Libsass come to this error message?
Second, the file begins with two lines of comment plus an extra return. If you take those out, the error magically appears exactly on `35:5`, but with those in it appears on `38:5`. Does LibSass discard leading whitespace? If so, this may (very likely) account for why the line numbers were reported wrong, and may account for some source map misalignment people are seeing with large projects (speculating on the last part though)

**Comment by mgreter** [comment]
@Snugug without seeing any examples we cannot really give any usefull comments on that! And no libsass does not discard any leading whitespace. Actually parsing position was never really a problem. Problem with source-maps stem from output (prefixing things like charset headings) and AST_Node manipulations (like in extend or eval)! So this is fixed by https://github.com/sass/libsass/pull/1062 then?

**Comment by Snugug** [comment]
@mgreter As linked above, the example is the `ms-merge` function [from Modular Scale](https://github.com/modularscale/modularscale-sass/blob/2.x/stylesheets/modular-scale/_sort-list.scss#L4-L42). Sorry if that link wasn't clear. Here's a copy of the function:

```scss
@function ms-merge($A, $B) {
  
  $Return: ();
  
  // Some empty lists get passed through
  // so just pass the other list throguh
  @if length($A) == 0 {
    @return $B;
  }

  // If lists fit next to each other, just merge them
  // This helps performance skipping the need to check each value
  @if nth($A, length($A)) < nth($B, 1) {
    @return join($A, $B);
  }
  @if nth($B, length($B)) < nth($A, 1) {
    @return join($B, $A);
  }
  
  // Counters start at 1
  $A-counter: 1;
  $B-counter: 1;
  
  // Start looping through all numbers in array
  @while $A-counter <= length($A) and $B-counter <= length($B) {
    
    // Check if the A value is smaller
    // Uses or equal to avoid duplicate numbers
    @if nth($A, $A-counter) <= nth($B, $B-counter) {
      $Return: join($Return, nth($A, $A-counter));
      $A-counter: $A-counter + 1;
    }
    
    // Check if the B value is smaller
    @elseif nth($A, $A-counter) > nth($B, $B-counter) {
      $Return: join($Return, nth($B, $B-counter));
      $B-counter: $B-counter + 1;
    }
  }
```

Trying to compile that as-is throws an error, changing the `@elseif` to `@else if` no longer throws that error.

**Comment by Snugug** [comment]
And yes, I think this specific instance will probably be solved by that, but maybe keep this open to figure out why the error line and message were wrong.

**Comment by mgreter** [comment]
@Snugug please don't take this wrong, but we prefer samples that can be run on their own without any guessing from our side. Your example above actually doesn't work since it is missing a closing bracket. It seems to compile correctly on when I add it on http://sassmeister.com/ (although it produces no output, but I guess that's ok for this report). Please open a new issue with reproducible steps for the error message issue you described. Thanks!

**Comment by Snugug** [comment]
@mgreter I must have copied it wrong from the source, now [linked three times](https://github.com/modularscale/modularscale-sass/blob/2.x/stylesheets/modular-scale/_sort-list.scss#L4-L62). Below is the correctly copied source:

```scss
@function ms-merge($A, $B) {
  
  $Return: ();
  
  // Some empty lists get passed through
  // so just pass the other list throguh
  @if length($A) == 0 {
    @return $B;
  }

  // If lists fit next to each other, just merge them
  // This helps performance skipping the need to check each value
  @if nth($A, length($A)) < nth($B, 1) {
    @return join($A, $B);
  }
  @if nth($B, length($B)) < nth($A, 1) {
    @return join($B, $A);
  }
  
  // Counters start at 1
  $A-counter: 1;
  $B-counter: 1;
  
  // Start looping through all numbers in array
  @while $A-counter <= length($A) and $B-counter <= length($B) {
    
    // Check if the A value is smaller
    // Uses or equal to avoid duplicate numbers
    @if nth($A, $A-counter) <= nth($B, $B-counter) {
      $Return: join($Return, nth($A, $A-counter));
      $A-counter: $A-counter + 1;
    }
    
    // Check if the B value is smaller
    @elseif nth($A, $A-counter) > nth($B, $B-counter) {
      $Return: join($Return, nth($B, $B-counter));
      $B-counter: $B-counter + 1;
    }
  }
  
  // Run through remainder values in the list
  @while $A-counter <= length($A) {
    $Current: nth($A, $A-counter);
    @if $Current != nth($Return, length($Return)) {
      $Return: join($Return, $Current);
    }
    $A-counter: $A-counter + 1;
  }
  @while $B-counter <= length($B) {
    $Current: nth($B, $B-counter);
    @if $Current != nth($Return, length($Return)) {
      $Return: join($Return, $Current);
    }
    $B-counter: $B-counter + 1;
  }
  
  // Done! return is now sorted and complete
  @return $Return;
}
```

**Comment by xzyfer** [comment]
FWIW the `@elsif` issue has since been fixed https://github.com/sass/libsass/issues/1060

**Comment by xzyfer** [comment]
Let me try answering your questions @Snugug 

>Changing the @elseif to @else if resolves this issue, although raises other question.

This was reported a couple hours ago #1060 and fixed #1062.

>First, how did Libsass come to this error message?

The nature of the previous bug meant that `@elseif` was not being recognised a control directives and hence was being treated as an `@at-directive` like `@media` or `@supports`. Given that the error message makes sense and is inline with Ruby Sass e.g. http://sassmeister.com/gist/1db38d074e50ed9697a8

>Second, the file begins with two lines of comment plus an extra return. If you take those out, the error magically appears exactly on 35:5, but with those in it appears on 38:5. 

This appears to be a real problem. For some reason error tracking is drifting. @mgreter 

***

This issue is resolved as of #1062, but there is a new issue in the incorrect error line reporting. 

If someone could kindly find the generated a sample of runnable Sass code via http://sassmeister.com/ that reproduces the incorrect error line and open the issue it'd be really appreciated.
